### DISCUSSION TITLE
Differentiate Between Obsolete Warnings

### DISCUSSION THREAD

**Discussion by MattBussing** [dotnet/runtime#91337] [Ideas] [answered]
# Background and Problem
At my job we have a very large c# code base. We have turned on all warnings as errors except for the obsolete warnings (CS0618 and CS0612). Since our code base is so large we can't clean up everything in one pull request, so we use the [ObsoleteAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-8.0) to tell developers to use a different method while we are cleaning up the old method. This blocks us from upgrading CS0618 and CS0612 to errors. 

There is a problem with this. Currently we have about 1000 obsolete warnings. When we go to upgrade a dependency all of these warnings hide new warnings added. For example, [here](https://stackoverflow.com/q/65847723/9728299). 

# Ideas
## 1 Custom Warning Codes
Update the [ObsoleteAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-8.0) to let us pass in a custom error code. Then, we can mark CS0618 and CS0612 as errors and set custom error codes for the methods we mark as obsolete but don't fix right away.
## 2 Warnings Baseline
Create a warnings baseline. Android has a concept called a [warnings baseline](https://developer.android.com/studio/write/lint#snapshot). 

> You can take a snapshot of your project's current set of warnings, then use the snapshot as a baseline for future inspection runs so that only new issues are reported. The baseline snapshot lets you start using lint to fail the build without having to go back and address all existing issues first.

eslint has a concept of [lint-todo](https://github.com/lint-todo/eslint-formatter-todo). 

Can you add something like this so that we can create a baseline of our current warnings? Then, when new warnings (or errors) are introduced, we'd be aware of them.

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
> Update the [ObsoleteAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-8.0) to let us pass in a custom error code.

The [ObsoleteAttribute.DiagnosticId](https://learn.microsoft.com/dotnet/api/system.obsoleteattribute.diagnosticid?view=net-5.0) property was added in .NET 5.

**Reply by MattBussing** [reply]
That is amazing! The problem is we have a lot of .NET Standard and .NET Framework code, so we can't use this until we upgrade to .NET 5. Anyone know of any plans to get this to .NET Standard or .NET Framework?

**Reply by KalleOlaviNiemitalo** [reply]
According to <https://learn.microsoft.com/en-us/dotnet/standard/net-standard>, "No new versions of .NET Standard will be released".

I don't know about future .NET Framework versions.  .NET Framework 4.8 was supposed to be the last version but then Microsoft published .NET Framework 4.8.1.

**Reply by MattBussing** [reply]
I was thinking maybe they'd add this code to a NuGet package so you could have access to it in .NET Standard or .NET Framework.

**Reply by KalleOlaviNiemitalo** [reply]
To let the compiler recognise the meaning of the attribute, it must be called System.ObsoleteAttribute.  A NuGet package would be able to define a class with that name but would not be able to remove or modify the class that is defined in mscorlib.  So any `[Obsolete("message")]` decoration would be ambiguous.  If the new ObsoleteAttribute were defined in the same project in which it is used, then the compiler would issue warning CS0436 and choose the local definition. But if the definition were in a referenced NuGet package, then I think it would be an error and solving it would require `extern alias`.  The transitive reference to this NuGet package would also cause similar ambiguity errors in projects that reference the project that has the obsolete types and members, and those projects would have to add direct references to the package in order to configure the `extern alias`.  Having two classes named System.ObsoleteAttribute would cause some difficulty in runtime reflection, too; software that searches for the attribute defined in mscorlib would not find the other attribute.

```csharp
using System;

namespace System {
    class ObsoleteAttribute : Attribute {
        public ObsoleteAttribute(string message) {}
        public int Buff { get; set; }
    }
}

public class C {
    [Obsolete("hah", Buff = 1)] // warning CS0436
    public void M() {
    }
    public void N() => M(); // warning CS0618, as intended
}
```

**Reply by MattBussing** [reply]
Good point.

**Comment by RobSwDev** [comment]
Can't you simply suppress all your existing Obsolete warnings in GlobalSuppressions.cs?
The Roslyn analyzer has an automatic fix, which you can apply solution-wide, to create your "warnings baseline".
Even on a large codebase, it shouldn't take too long. 
Then you can make the obsolete warning an error again.

![image](https://github.com/dotnet/msbuild/assets/26545948/44a80299-a726-48b6-9f3a-0d50a5cf90d5)

**Reply by MattBussing** [reply]
I'm going to have to try this. Thank you!

**Reply by MattBussing** [reply]
I wasn't able to get this working for CS0618. :(
